src/regex.js

```diff
@@ -0,0 +1,53 @@
+import _ from 'lodash/fp'
+
+export const testRegex = regex => regex.test.bind(regex)
+export const makeRegex = options => text => RegExp(text, options)
+export const makeAndTest = options => _.flow(makeRegex(options), testRegex)
+export const matchAnyWord = _.flow(
+  _.words,
+  _.map(makeAndTest('gi')),
+  _.overSome
+)
+
+export const postings = (regex, str) => {
+  var match = regex.exec(str)
+  let result = []
+
+  while (match) {
+    result.push([match.index, regex.lastIndex])
+    match = regex.exec(str)
+  }
+  return result
+}
+
+let push = (val, arr) => arr.concat([val])
+export const postingsForWords = (string, str) => _.reduce(
+  (result, word) => push(postings(RegExp(word, 'gi'), str), result), []
+)(_.words(string))
+
+export const insertAtIndex = (index, val, str) => str.slice(0, index) + val + str.slice(index)
+
+export const mergeRanges = (p1, p2) => ((p2[0] <= p1[1]) && [[p1[0], _.max([p1[1], p2[1]])]]) || [p1, p2]
+
+const mergeRangesOver = (result, posting) => {
```

Comment:
what we really want is just a version of mergeRanges that can take more than two ranges